consumer.go

```diff
@@ -118,18 +118,18 @@ func (c *Consumer) Scan(ctx context.Context, fn ScanFunc) error {
 
 	wg := new(sync.WaitGroup)
 	// process each of the shards
-	shardsInProcess := make(map[string]struct{})
+	s := newShardsInProcess()
 	for shard := range shardC {
 		shardId := aws.ToString(shard.ShardId)
-		if _, ok := shardsInProcess[shardId]; ok {
+		if s.doesShardExist(shardId) {
 			// safetynet: if shard already in process by another goroutine, just skipping the request
 			continue
 		}
 		wg.Add(1)
 		go func(shardID string) {
-			shardsInProcess[shardID] = struct{}{}
+			s.addShard(shardID)
 			defer func() {
-				delete(shardsInProcess, shardID)
+				s.deleteShard(shardID)
```

Comment:
wonder if `removeShard` or `forgetShard` could be a little softer; the `deleteShard` reads like we should be deleting something vs removing it